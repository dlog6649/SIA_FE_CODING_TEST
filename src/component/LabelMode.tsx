import React, { useState, useEffect, useRef } from "react";
import { LABEL_CREATE_MODE, LABEL_SELECT_MODE } from "../modules/annotator/types";
import imgLabelSelectMode from "../asset/images/label_select_mode.png";
import imgLabelCreateMode from "../asset/images/label_create_mode.png";

interface Props {
  mode: string;
  changeMode: (mode: string) => void;
}

export default function LabelMode(props: Props) {
  const [mode, setMode] = useState<string>(props.mode);
  const refModeBtnList = useRef<HTMLDivElement>(null);

  useEffect(() => {
    console.log("LabelMode useEffect: [props.mode]");
    if (!refModeBtnList.current) {
      return;
    }
    const refModeBtnDiv = refModeBtnList.current;
    refModeBtnDiv.childNodes.forEach((modeBtnEle) => {
      const modeBtn = modeBtnEle as HTMLButtonElement;
      props.mode === modeBtn.id ? modeBtn.classList.add("active") : modeBtn.classList.remove("active");
    });
  }, [props.mode]);

  const clickBtn = (evt: React.FormEvent<HTMLButtonElement>) => {
    const clickedMode: string = evt.currentTarget.id;
    if (mode === clickedMode) {
      return;
    }
    if (!refModeBtnList.current) {
      return;
    }
    const refModeBtnDiv = refModeBtnList.current;
    refModeBtnDiv.childNodes.forEach((modeBtnEle) => {
      const modeBtn = modeBtnEle as HTMLButtonElement;
      if (clickedMode === modeBtn.id) {
        modeBtn.classList.add("active");
      } else {
        modeBtn.classList.remove("active");
      }
    });
    setMode(clickedMode);
    props.changeMode(clickedMode);
  };

  return (
    <div className="label-mode" ref={refModeBtnList}>
      <button id={LABEL_SELECT_MODE} className="btn label-mode-btn active" type="button" onClick={clickBtn}>
        <img className="btn-img" src={imgLabelSelectMode} alt="label_select_mode" />
      </button>
      <button id={LABEL_CREATE_MODE} className="btn label-mode-btn" type="button" onClick={clickBtn}>
        <img className="btn-img" src={imgLabelCreateMode} alt="label_create_mode" />
      </button>
    </div>
  );
}
