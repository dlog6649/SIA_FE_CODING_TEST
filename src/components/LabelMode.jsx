import React from 'react';

export default function LabelMode(props) {

    const clickBtn = e => {
        let mode;

        document.querySelector('.label-mode').childNodes.forEach(node => {
            if(e.currentTarget.id === node.id) {
                node.classList.add('active');
                mode = node.id;
            }
            else {
                node.classList.remove('active');
            }
        })
        
        props.onClick(mode);
    }

    return (
        <div className="label-mode">
            <button id="LABEL_SELECT_MODE" className="btn label-mode-btn active" type="button" onClick={clickBtn}>
                <img className="btn-img" src={require('../images/label_select_mode.png')} alt="label_select_mode"/>
            </button>
            <button id="LEBEL_CREATE_MODE" className="btn label-mode-btn" type="button" onClick={clickBtn}>
                <img className="btn-img" src={require('../images/label_create_mode.png')} alt="label_create_mode"/>
            </button>
        </div>
    );
}
