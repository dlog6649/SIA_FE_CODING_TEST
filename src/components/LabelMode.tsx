import React, { useState, useEffect, useRef } from 'react';
import { LABEL_CREATE_MODE, LABEL_SELECT_MODE } from '../modules/annotator';

interface Props {
    mode: string
    changeMode: (mode: string) => void
}

export default function LabelMode(props: Props) {
    const [mode, setMode] = useState(props.mode);
    const refModeBtnList: React.MutableRefObject<any> = useRef(null);

    useEffect(() => {
        console.log('LabelMode useEffect: [props.mode]');
        refModeBtnList.current.childNodes.forEach((modeBtn: HTMLElement) => {
            props.mode === modeBtn.id ? modeBtn.classList.add('active') : modeBtn.classList.remove('active');
        });
    }, [props.mode]);

    const clickBtn = (evt: React.FormEvent<HTMLButtonElement>) => {
        const clickedMode: string = evt.currentTarget.id;
        if(mode === clickedMode) {
            return;
        }
        refModeBtnList.current.childNodes.forEach((modeBtn: HTMLElement) => {
            if(clickedMode === modeBtn.id) {
                modeBtn.classList.add('active');
            }
            else {
                modeBtn.classList.remove('active');
            }
        });
        setMode(clickedMode);
        props.changeMode(clickedMode);
    }
    
    return (
        <div className="label-mode" ref={refModeBtnList}>
            <button id={LABEL_SELECT_MODE} className="btn label-mode-btn active" type="button" onClick={clickBtn}>
                <img className="btn-img" src={require('../asset/images/label_select_mode.png')} alt="label_select_mode"/>
            </button>
            <button id={LABEL_CREATE_MODE} className="btn label-mode-btn" type="button" onClick={clickBtn}>
                <img className="btn-img" src={require('../asset/images/label_create_mode.png')} alt="label_create_mode"/>
            </button>
        </div>
    );
}
