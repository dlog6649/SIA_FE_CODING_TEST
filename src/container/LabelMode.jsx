import LabelMode from '../components/LabelMode';
import { connect } from 'react-redux';
import { changeMode } from '../modules/annotator';


const mapStateToProps = state => {
    return {
        mode: state.annotator.mode
    };
}


const mapDispatchToProps = dispatch => {
    return {
        changeMode: mode => dispatch(changeMode(mode))
    };
}


export default connect(mapStateToProps, mapDispatchToProps)(LabelMode);