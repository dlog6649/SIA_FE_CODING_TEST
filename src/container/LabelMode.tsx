import { connect } from "react-redux";
import LabelMode from "../components/LabelMode";
import { changeMode, AnnotatorState } from "../modules/annotator";

interface State {
  annotator: AnnotatorState;
}

const mapStateToProps = (state: State) => {
  return {
    mode: state.annotator.mode,
  };
};

const mapDispatchToProps = (dispatch: any) => {
  return {
    changeMode: (mode: string) => dispatch(changeMode(mode)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(LabelMode);
