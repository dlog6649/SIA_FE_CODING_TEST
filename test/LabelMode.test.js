import React from "react";
import { render, fireEvent } from "@testing-library/react";
import LabelMode from "./LabelMode";

describe("LabelMode 테스트 시작", () => {
  let wrapper = null;
  const mockChangeMode = jest.fn();
  const _mode = "LABEL_SELECT_MODE";

  it("렌더링된 스냅샷이 기존과 일치해야 한다.", () => {
    wrapper = render(<LabelMode mode={_mode} changeMode={mockChangeMode} />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it("라벨모드 버튼을 클릭 시 활성화되며, 해당 모드는 디스패치돼야 한다.", () => {
    wrapper = render(<LabelMode mode={_mode} changeMode={mockChangeMode} />);
    const buttons = wrapper.getAllByRole("button");

    // 생성모드 버튼 클릭
    fireEvent.click(buttons[1]);

    expect(buttons[0].classList.contains("active")).toBe(false);
    expect(buttons[1].classList.contains("active")).toBe(true);

    const changeModeEventByClick = mockChangeMode.mock.calls;
    const [mode] = changeModeEventByClick[0];
    expect(mode).toEqual("LABEL_CREATE_MODE");
  });
});
