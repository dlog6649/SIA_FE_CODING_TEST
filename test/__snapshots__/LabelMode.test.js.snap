// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelMode 테스트 시작 렌더링된 스냅샷이 기존과 일치해야 한다. 1`] = `
<body>
  <div>
    <div
      class="label-mode"
    >
      <button
        class="btn label-mode-btn active"
        id="LabelMode.SELECT"
        type="button"
      >
        <img
          alt="LabelMode.SELECT"
          class="btn-img"
          src="LabelMode.SELECT.png"
        />
      </button>
      <button
        class="btn label-mode-btn"
        id="LabelMode.CREATE"
        type="button"
      >
        <img
          alt="LabelMode.CREATE"
          class="btn-img"
          src="LabelMode.CREATE.png"
        />
      </button>
    </div>
  </div>
</body>
`;
